File: Leetcode/993. Cousins in Binary Tree.py
class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        xDepth = yDepth = -1
        xPar = yPar = None

        def levelOrder(node):
            nonlocal xDepth, yDepth, xPar, yPar
            if not node:
                return
            q = []
            level = 0
            q.append((node, None))
            while q:
                n = len(q)
                for i in range(n):
                    tmp, parent = q.pop(0)
                    if tmp.val == x:
                        xDepth = level
                        xPar = parent
                    if tmp.val == y:
                        yDepth = level
                        yPar = parent

                    if tmp.left: q.append((tmp.left, tmp))
                    if tmp.right: q.append((tmp.right, tmp))

                level += 1

        levelOrder(root)
        return xDepth == yDepth and xPar != yPar

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: Optional[TreeNode], x: int, y: int) -> bool:
        info = {}
        def levelOrder(node):
            if not node:
                return
            q = []
            level = 0
            q.append((node, None))
            while q:
                levelSize = len(q)
                for i in range(levelSize):
                    tmp, parent = q.pop(0)
                    info[tmp.val] = (level, parent)
                    if (tmp.left): q.append((tmp.left, tmp))
                    if (tmp.right): q.append((tmp.right, tmp))
                level += 1

        levelOrder(root)
        # print(info)
        
        return info[x][0] == info[y][0] and info[x][1] != info[y][1]

File: Leetcode/993. Cousins in Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isCousins(TreeNode* root, int x, int y) {
        queue<pair<TreeNode*, TreeNode*>> q;
        if (!root) return false;

        q.push(make_pair(root, nullptr));
        while (!q.empty()) {
            int n = q.size();
            TreeNode *xParent = nullptr, *yParent = nullptr;
            for (int i = 0; i < n; i++) {
                TreeNode* tmp = q.front().first;
                TreeNode* parent = q.front().second;
                q.pop();

                if (tmp->val == x) xParent = parent;
                if (tmp->val == y) yParent = parent;

                if (tmp->left) q.push(make_pair(tmp->left, tmp));
                if (tmp->right) q.push(make_pair(tmp->right, tmp));
            }
            if (xParent && yParent) return xParent != yParent;
            if ((xParent && !yParent) || (!xParent && yParent)) return false;
        }
        return false;
    }
};
